### PR TITLE
Add TextColor property to BaseButton and make MathInputButton's color customizable

### PR DESCRIPTION
**Pull request by SymboLinker**
The main goal of this commit is to make keyboard colors customizable (per keyboard key) to fit an app's color theme(s). For example, a dark keyboard background with light button texts.

CSharpMath.Forms:
- Refactor BaseButton's constructor.
- Add TextColor property to BaseButton and a bindable TextColorProperty.
- Remove the predefined Red color from MathInputButton.InputToLaTeX(input) for "Backspace" and "Clear".

CSharpMath.Forms.Example (MathKeyboard.xaml):
- Use the TextColorProperty in the Style of "Backspace" and "Clear", setting it to Red.

### CODE REVIEW

**Comment by Happypig375** [comment]
Not a fan of use-once functions. @FoggyFinder?

**Comment by Happypig375** [comment]
Also would be nice if some new tests related to this change are added to CSharpMath.Xaml.Tests.

**Comment by FoggyFinder** [comment]
> Not a fan of use-once functions. @FoggyFinder?

Yep, as well

**Comment by SymboLinker** [comment]
OK, I will restore that and make some unit tests.

Op wo 16 sep. 2020 09:15 schreef FoggyFinder <notifications@github.com>:

File: CSharpMath.Forms/Buttons.cs
Comment (Happypig375) [inline CSharpMath.Forms/Buttons.cs:36]: This function can be inlined into the property setter as well by initializing `TextColor` in the constructor.

File: CSharpMath.Forms/Buttons.cs
Comment (Happypig375) [inline CSharpMath.Forms/Buttons.cs:58]: Can be condensed into `=> ((BaseButton<TView, TPainter, TContent>)b).TextColor = (Color)n`.

**Comment by SymboLinker** [comment]
Oops, I didn't read "CSharpMath.Xaml.Tests". 

While trying to find out where to write the unit tests, I was in doubt about the "Xaml" meaning. I thought that test project was meant for tests more related to the xml. I have created a test project "CSharpMath.Forms.Tests" in the "Xamarin.Forms" folder. Shall I move the tests and delete the newly created test project?

File: CSharpMath.Forms.Tests/ButtonTests.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/ButtonTests.cs:37]: I avoid using the null suppression operator. Should we allow it in tests or define a `NotNull` function which will raise `Xunit.Sdk.NotNullException` instead?

File: CSharpMath.Forms/LatexHelper.cs
Comment (Happypig375) [inline CSharpMath.Forms/LatexHelper.cs:8]: `\hphantom` is for horizontal spacing while `\vphantom` is for vertical spacing. So `hphantom` would be more appropriate here.

**Review by Happypig375 — COMMENTED** [review]
Hmmm, if you create a new test project then make sure to include it in `CSharpMath.CrossPlatform.slnf` as well.

File: CSharpMath.Forms/LatexHelper.cs
Comment (SymboLinker) [reply CSharpMath.Forms/LatexHelper.cs:8]: The main purpose of the temporary "phantom" is, if I'm correct, to make sure that small symbols like the degree symbol are not streched. Using the vertical "|" symbol prevents that from happening. The comment in the code also says something about horizontal spacing, but I think that is not its main purpose. Besides, spacing between buttons can also be created in other ways that may be more appropriate - maybe the comment in the code is not correct.

File: CSharpMath.CrossPlatform.slnf
Comment (Happypig375) [inline CSharpMath.CrossPlatform.slnf:32]:
```suggestion
	"CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj"
```

File: CSharpMath.Forms.Tests/ButtonTests.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/ButtonTests.cs]:
```suggestion
    public void AllMathInputButtonsHaveLatexContent(MathInputButton mathInputButton) =>
      Assert.False(string.IsNullOrEmpty(mathInputButton.Content?.LaTeX));
```

File: CSharpMath.Forms.Tests/ButtonTests.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/ButtonTests.cs:41]:
```suggestion
    public void MathInputButtonsHaveTransparentBackgroundByDefault(MathInputButton mathInputButton) =>
      Assert.Equal(Color.Transparent, mathInputButton.BackgroundColor);
```

File: CSharpMath.Forms.Tests/TestHelpers/ComplexClassData.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/TestHelpers/ComplexClassData.cs:3]: This is just duplicating TheoryData of Xunit.

File: CSharpMath.Forms.Tests/ButtonTests.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/ButtonTests.cs:42]: I suggest `MemberData` instead of `ClassData` for less code.

File: CSharpMath.Forms.Tests/TestHelpers/NotNull.cs
Comment (Happypig375) [inline CSharpMath.Forms.Tests/TestHelpers/NotNull.cs:12]: Why are the pragma directives here?

File: CSharpMath.Forms/LatexHelper.cs
Comment (Happypig375) [reply CSharpMath.Forms/LatexHelper.cs:8]: Hmmm. What do you suggest on horizontal spacing?

File: CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj
Comment (Happypig375) [inline CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj:3]: If you're not testing adding custom typefaces, then you should omit this line.

File: CSharpMath.Forms.Tests/ButtonTests.cs
Comment (Happypig375) [reply CSharpMath.Forms.Tests/ButtonTests.cs:42]: Also, should use `MathKeyboardInput` as parameters since only xunit-serializable types will result in individual test cases instead of one single test case with multiple results.

File: CSharpMath.Forms/LatexHelper.cs
Comment (SymboLinker) [reply CSharpMath.Forms/LatexHelper.cs:8]: I did not know about the \phantom command before working on this pull request, and I learned that the \phantom is a square (and \vphantom is a vertical line of zero width). If the \phantom is implemented and used instead of "|", the horizontal space will be increased. If a \vphantom is implemented, yes, then horizontal spacing must be managed another way. I think that managing the width of the button should be done by the developers using the buttons. Probably they will define the width of the buttons as a percentage of the screenwidth or something like that anyway. But if the intention is  "providing a helping hand in case the width is not managed by the developers", then probably the simple solution is the right one and using an invisible "|" is exactly that: it is related to the font size, which seems just perfect.
I have tried to find examples however where that minimum width is used. Is there an example in the example project that uses the minimum width that "|" causes?

Comment (SymboLinker) [reply CSharpMath.Forms/LatexHelper.cs:8]: Leaving the "|" width both responsibilities may be better than using \phantom or \vphantom. Developers that define width themselves don't notice it, and developers that do not define width are helped without asking. In that case the comment can be updated (remove the line that mentions \phantom).

Comment (Happypig375) [reply CSharpMath.Forms/LatexHelper.cs:8]: See the Editor tab in CSharpMath.Forms.Example, I find that adding a tiny bit of horizontal space makes it look better

Comment (Happypig375) [reply CSharpMath.Forms/LatexHelper.cs:8]: Well using `\phantom` is really the intention here. `\color` with transparent is a hack.

File: CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj
Comment (SymboLinker) [reply CSharpMath.Forms.Tests/CSharpMath.Forms.Tests.csproj:3]: After deleting that line at least one of the unit tests failed. Maybe that was not the only thing I changed. I will re-check. (I thought it was related to creating the stream for setting the ImageSource of BaseButton and did not think about it really, I'm sorry.)

File: CSharpMath.Forms.Tests/TestHelpers/ComplexClassData.cs
Comment (SymboLinker) [reply CSharpMath.Forms.Tests/TestHelpers/ComplexClassData.cs:3]: O, nice. I will delete this class and replace usings of it.

**Comment by SymboLinker** [comment]
I really appreciate the feedback you are providing. This is the first pull request I've ever done, and I am surprised about the quick responses and thorough reviews, providing kind and helpful suggestions to make the code compact and clean. You also care about the code quality even more than I'd expected - commenting about a white line for example and using `=> bla;` instead of `=> { bla;}` - and I like that.

**Review by Happypig375 — COMMENTED** [review]
It's because I am a victim of dead air in pull requests...
https://github.com/kostub/iosMath/pull/142
https://github.com/LayoutFarm/Typography/pull/192
https://github.com/fsharp/fslang-design/pull/442
https://github.com/MicrosoftDocs/visualfsharpdocs/pull/326
And you will most definitely face similar scenarios as well in the future.

Regarding code style, thanks - less code means fewer chances of bugs.

**Comment by SymboLinker** [comment]
Maybe I should only have committed "Add MathInputButton_Command unit test and do the suggested cleaning 99417ce".
I thought "let's use the Keyboard getter to increase test coverage", but the intention of the test is more clear without it. I will revert the last two commits.

**Comment by Happypig375** [comment]
Try having tests that invoke https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.streamimagesource.stream?view=xamarin-forms#Xamarin_Forms_StreamImageSource_Stream of the `Source` property.

**Comment by SymboLinker** [comment]
Hmm. The new tests fail in the github checks (not locally)...:

> The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.

**Comment by Happypig375** [comment]
CSharpMath.Rendering.Tests test the SkiaSharp front end yet all images are drawn successfully. Weird.

**Comment by Happypig375** [comment]
Version mismatch? SkiaSharp is at 2.80.1 while SkiaSharp.Views.Forms is at 1.68.0.

**Comment by SymboLinker** [comment]
When I update the package, another error shows up, which is also mentioned in https://github.com/mono/SkiaSharp/issues/1225.
A comment in that issue:

> It might be because the SkiaSharp.Views.Forms project does not actually have a netstandard implementation that can run in tests.

**Comment by SymboLinker** [comment]
I've found a solution for using 2.80.2. I've downloaded the NuGetPackage, extracted it (zip file) and added the dll to the test project:
skiasharp.views.forms.2.80.2.nupkg\ref\netstandard2.0\SkiaSharp.Views.Forms.dll

I'll do some more checks and then I'll push.

**Comment by SymboLinker** [comment]
Just clarifying: I've had two versions of the repository (with SkiaSharp.Views.Forms 1.68.0 and 2.80.2) that run the tests of CSharpMath.Forms.Tests successfully locally but fail in the GitHub checks.
I now continue my quest :)

**Comment by SymboLinker** [comment]
OK, that didn't help.

Github Test  Error Message:
   System.AggregateException : One or more errors occurred. (The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.)
---- System.TypeInitializationException : The type initializer for 'SkiaSharp.SKImageInfo' threw an exception.
-------- **System.DllNotFoundException : Unable to load shared library 'libSkiaSharp' or one of its dependencies.** In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: liblibSkiaSharp: cannot open shared object file: No such file or directory

**Comment by SymboLinker** [comment]
Solving this NuGet/dll problem seems out of reach for me. I will revert my latest attempts. Options:
- We could remove or comment out the unit tests added in "Add MathButton unit tests that check that the image color is correcty set in the StreamImageSource (4223835)" and wait for the mono/SkiaSharp update to solve the bug (assuming that is the case).
-  I don't know on what device Github Tests run? I found a posting about configuring the machine on which the tests are run (I can't help any further with that...)
https://github.com/mono/SkiaSharp/issues/834#issuecomment-631479750
- Someone may be able to solve it another way...

**Comment by Happypig375** [comment]
> I don't know on what device Github Tests run?

The workflow runs on Linux (Ubuntu).

**Comment by Happypig375** [comment]
I'll follow up tomorrow, I think https://github.com/mono/SkiaSharp/releases/tag/v2.80.0#additional should be the solution.

**Comment by Happypig375** [comment]
I'm busy these few days, will come back to this a few more days later.

**Comment by SymboLinker** [comment]
I will push another commit that will allow/stimulate customizing the keyboard keys even more.

from
image

to (for example)
image

**Comment by Happypig375** [comment]
One of the squares is filled because it indicates where the cursor will be after pressing the button.

**Comment by SymboLinker** [comment]
The main reason for not liking an empty square is that when I make the keyboard smaller vertically, the displayed borders of some empty sqaures are not of equal size. That may because of my computer screen, but still a reason to like to override the default:
image

Shall I remove the changes of the square in the Example project?

Or: make the empty square gray and leave the black square, via something like (using lumosity in case of TextColor changes):
```
  public class MyMathInputButton : MathInputButton {
    protected override string GetButtonDisplayLaTeX(MathKeyboardInput key) =>
      base.GetButtonDisplayLaTeX(key).Replace(@"\square", LatexHelper.SetColor("■", TextColor.WithLuminosity(0.7)));
    }
  }
```
image

**Comment by Happypig375** [comment]
If the button display is changed, the keyboard output should be changed too. @charlesroddie @FoggyFinder Is this change acceptable?

**Comment by SymboLinker** [comment]
I would say that if the button display is customizable, the keyboard output need not change.
The keyboard output could become customizable too (that would be great actually), but I think those two need not be always the same.
- Imagine that someone would like to display `\times` on a button, but `\cdot` as output.
- Imagine that someone wants to make a calculator with `x^y` as the button-LaTeX...

**Comment by FoggyFinder** [comment]
> If the button display is changed, the keyboard output should be changed too. @charlesroddie @FoggyFinder Is this change acceptable?

I think it would be better to move these changes into separate PR (or create an issue to discuss first).

But I personally don't like how new buttons look.

**Comment by Happypig375** [comment]
Personally I am okay with the fire button but don't like the idea of what you see not being what you get.

**Comment by SymboLinker** [comment]
Just to clarify:
The point of discussion is whether MathInputButton.GetButtonDisplayLaTeX should be virtual or not.

In a discussion about that, the following comments...
- "But I personally don't like how new buttons look."
- "don't like the idea of what you see not being what you get."
...seem irrelevant because everyone can get what they want. The defaults stay the same.

A concern that is more important in my opinion:
Does this change have any negative technical implications for the development of this repository? For example: does it impose restrictions for classes beyond the MathInputButton class? (Does it violate the open–closed principle or does it make future changes harder because GetButtonDisplayLaTeX is not private anymore?)

I must say however, that this comment
"If the button display is changed, the keyboard output should be changed too."
is another example of feedback that I really appreciate. I'm just someone opening a pull request, and a proposal is taken [more serious than expected] / [beyond the intended scope].

**Comment by SymboLinker** [comment]
For making the placeholder customizable in the output, I have created pull request https://github.com/verybadcat/CSharpMath/pull/167.

**Comment by SymboLinker** [comment]
My proposal/plan for this PR after #167 has been merged/accepted:
- I'll sync/update this branch.
- I'll restore the placeholder appearances on the keyboard keys to what they were before.
- I'll add a button at the right of the Textbox "Enter keystrokes..." that says something like "Change appearance". Maybe only a single word is better, to keep more space for the textbox.
- The effect of clicking this new button that has the following effect **the first time**: the placeholder appearance changes on the keyboard buttons and in the output as well. (For example, using a LightGray full square instead of the currently used empty square with black border.)
- Clicking multiple times changes the appearance of keyboard keys again, for instance also changing the TextColor of all keys to brown, or changing the Clear button to a rubbish bin, and eventually (after a maximum of 5 clicks) the original appearance is visible again. This way, the defaults stay the same, but all added functionality is also available in the Example project.

**Comment by SymboLinker** [comment]
I wanted to sync my fork in such a way that this pull request would not show the commit from #167, but alas. I chose "rebase", maybe "merge" would have done the trick. However, I also commited ddb5e4e and that is part of this pull request.

**Comment by Happypig375** [comment]
I'll try to solve Linux failures when I can allocate time...

**Comment by SymboLinker** [comment]
I understand. Sometimes it's juggling priorities. I will stay tuned.

**Comment by SymboLinker** [comment]
I would like to move on to other PRs but I also intend to avoid merge conflicts, so I'll wait with opening new PRs until this PR has been merged.

Proposal (I'll just do it and it can always be reverted):
- I remove the following two tests that have the Linux failures (DefaultButtonImageTextColorIsBlack, ButtonTextColorPropertyChangesImageColor).
- I will open a PR that re-adds the tests after this PR has been merged.

**Comment by charlesroddie** [comment]
`[Fact(Skip="reason")]` ?

**Review by FoggyFinder — CHANGES_REQUESTED** [review]
`TextColor` on buttons isn't reverted to `Black` after theme changed back to the first one.

**Comment by SymboLinker** [comment]
> TextColor on buttons isn't reverted to Black after theme changed back to the first one.

This has never happened while I tested, but there is definitely something strange going on: if I deploy/run the UWP project several times and each time do the "Change appearance" test through all themes, for some deploys 1 or 2 buttons do not revert to Black when coming back to the first theme, and in other cases it just works. Testing now, I had the "+" button that stayed brown, in another deploy I had the "5" and "\leq" buttons that stayed brown...

Something to admit: I have not tested this for Android, because that has a compilation error:

> Can not resolve reference: `Validation`, referenced by `Xunit.SkippableFact`. Please add a NuGet package or assembly reference for `Validation`, or remove the reference to `Xunit.SkippableFact`.	CSharpMath.Forms.Example.Android

I don't know why "xunit"-related stuff (I thought they are for unit test only) can be found in Android's obj/project.assets.json and at the NuGet packages while they are not referenced in Android's csproj:
afbeelding

**Comment by SymboLinker** [comment]
My latest commit does not fix the issue FoggyFinder described.

Still investigating (at some moments I have time).

**Comment by SymboLinker** [comment]
Ow, the unit tests think differently about removing the NullableColorBindablePropertyHelper. I will revert.

>     Xamarin.Forms.Xaml.XamlParseException : Position 1:97. Cannot assign property "PlaceholderActiveColor": Property does not exist, or is not assignable, or mismatching type between value and property

File: CSharpMath.Forms.Example/CSharpMath.Forms.Example/Controls/MathKeyboard.xaml.cs
Comment (FoggyFinder) [inline CSharpMath.Forms.Example/CSharpMath.Forms.Example/Controls/MathKeyboard.xaml.cs:59]:
```suggestion
                            .OfType<ImageSourceMathInputButton>()
```

**Comment by SymboLinker** [comment]
@FoggyFinder I can't reproduce the bug anymore (of MathInputButtons not changing back to black when coming back to the first theme). I have tested this now, about 10 times, and 15 hours ago, about 20 times. Does it still occur for you?

If it does, the only thing I can think of to change, is listening to the following warning:

image

by defining `interface IButtonDraw { void ButtonDraw(); }` and using it on BaseButton's TextColor property:
```
    public Color TextColor { get => (Color)GetValue(TextColorProperty); set => SetValue(TextColorProperty, value); }
    public static readonly BindableProperty TextColorProperty = BindableProperty.Create(nameof(TextColor), typeof(Color), typeof(BaseButton<TView, TPainter, TContent>), Color.Black,
      propertyChanged: (b, o, n) => ((IButtonDraw)b).ButtonDraw());
```
(Casting b to BaseButton<TView, TPainter, TContent> would not work, because MathInputButton's override of the ButtonDraw method would not be executed.)

**Comment by FoggyFinder** [comment]
> Does it still occur for you?

Yep, still the same

**Comment by SymboLinker** [comment]
Core Tests results:
> Total tests: 937
     Passed: 933
    Skipped: 4

The test fail errors are not related to this PR.

Test Core fail reason:

> Error: Codecov failed with the following error: The process '/bin/bash' failed with exit code 2

Test iOS fail reason:

> 2020-11-01 08:20:56.474 CSharpMath.Ios.Tests[2654:25586] 	[FAIL] CSharpMath.Ios.Tests.Tests.MathDisplay(file: "ItalicAlignment", latex: "\\colorbox{yellow}P\\\\\\begin{array}{r}\\colorbox"...) : Assert.InRange() Failure\r\nRange:  (-1655.8 - 6525.8)\r\nActual: 6688
2020-11-01 08:20:56.474 CSharpMath.Ios.Tests[2654:25586] 		  at CSharpMath.Ios.Tests.Tests.Test (System.String directory, System.Action`1[T] init, System.String file, System.String latex) [0x0020a] in /Users/runner/work/CSharpMath/CSharpMath/CSharpMath.Ios.Tests/Tests.cs:54 

> 2020-11-01 08:20:55.897 CSharpMath.Ios.Tests[2654:25586] 	[FAIL] CSharpMath.Ios.Tests.Tests.MathInline(file: "ItalicAlignment", latex: "\\colorbox{yellow}P\\\\\\begin{array}{r}\\colorbox"...) : Assert.InRange() Failure\r\nRange:  (-1655.8 - 6525.8)\r\nActual: 6688
2020-11-01 08:20:55.897 CSharpMath.Ios.Tests[2654:25586] 		  at CSharpMath.Ios.Tests.Tests.Test (System.String directory, System.Action`1[T] init, System.String file, System.String latex) [0x0020a] in /Users/runner/work/CSharpMath/CSharpMath/CSharpMath.Ios.Tests/Tests.cs:54

**Comment by SymboLinker** [comment]
@FoggyFinder I suspect that it is solved now via commit 486ac0f and I hope that calling ButtonDraw during each theme change in the example project is acceptable. Officially it is only needed to go from theme 1 to theme 2.

**Comment by Happypig375** [comment]
@FoggyFinder @SymboLinker I'll test this PR tomorrow

**Comment by SymboLinker** [comment]
> @FoggyFinder @SymboLinker I'll test this PR tomorrow

@Happypig375 It's more than half a year later now. Could you please share why this PR has not been tested / has not gotten any response? I would like to know that, so that I can do better next time (in other public repositories).

**Comment by Happypig375** [comment]
It's just me focusing on other projects than this. I'll just merge this.

**Review by Happypig375 — APPROVED** [review]

**Comment by Happypig375** [comment]
@FoggyFinder Just missing your approval

**Review by FoggyFinder — CHANGES_REQUESTED** [review]
Color doesn't return back to Black in theme switching. I'll take a look soon and write my thoughts / questions

**Comment by FoggyFinder** [comment]
> It's more than half a year later now. Could you please share why this PR has not been tested / has not gotten any response? I would like to know that, so that I can do better next time (in other public repositories).

I'm doing some testing. I totally newbie with Regex so getting help with those who know at least something (pretty much to my opinion). Expect feedback / commit with fix proposal or. at least. tests that are failing within 1-2 days.

**Comment by SymboLinker** [comment]
@FoggyFinder 
- Why do you need RegEx? Is it because of the use of `Regex.Matches(latex, LaTeXSettings.PlaceholderActiveNucleus).Count > 1` in the MathInputButton? That is acutally a bug, I realize now! (The intention of that piece of code was to count the number of placeholders. I actually should not have used the Regex class there, because my intention was only to count the number of occurrences of a normal string. That's dangerous however, because LaTeX uses characters that may result in actual meaningful regular expressions. It was lazy to use that method instead of writing a simple "substring counting method" - I believe there are no such methods ready available.)
- Whenever I "think" I need to use regular expressions, I remind myself of [a joke about Regex](): _"Some people, when confronted with a problem, think "I know, I'll use regular expressions." Now they have two problems."_ (I try to avoid using Regex. Often, there are better/cleaner ways to solve problems without Regex.)
- If I do use it, I mostly Google a "C# Regex cheat sheet" (for example: [mikesdotnetting](https://www.mikesdotnetting.com/article/46/c-regular-expressions-cheat-sheet)). There are multiple versions of Regex. Some languages outside of .NET may have a slightly different syntax.
- Regex.Replace and Regex.Matches are examples of .NET methods that you can use to transform those Regex strings into doing something.

**Comment by FoggyFinder** [comment]
Exactly. I thought "Aha. Now I realize why it behaves like this. And then - I can write simple Regex to fix it quickly. Well it ended with debug why a regex doesn't work properly in some case and then another regex and so on until the point when I totally no idea what the regex pattern means itself". 

Indeed I prefer not to use Regex and not only I'm very bad with them. But...you've been waiting for so long...my intention was to fix it ASAP,  merge in and then, probably, rewrite without Regex.

**Comment by FoggyFinder** [comment]
The issue with display (like I saw brown color despite the theme is black) is the fact that `DrawButton` calls repeatedly in short period of time:

```csharp
        button.TextColor = color;
        button.PlaceholderRestingColor = placeholderRestingColor;
        button.PlaceholderActiveColor = placeholderActiveColor;
        button.ButtonDraw(/* If the LaTeXSettings change but the button's appearance properties don't, there's no event that causes the execution of this method. */);
```

So when we store "original" latex in a variable

        var originalLatexString = painter.LaTeX;

it might be previously mutated LaTeX that actually contains color etc attributes.

**Comment by FoggyFinder** [comment]
I'm not a big fan of lock either but it was much easier than applying odd regex to prevent latex corruption.
Well, returning to black theme still shows light-red border in `frac`, ... but no visual glitch

**Comment by SymboLinker** [comment]
Nice, using a lock in synchronous code is new for me. Thanks!

File: CSharpMath.Forms/MathInputButton.cs
Comment (FoggyFinder) [inline CSharpMath.Forms/MathInputButton.cs:48]: One minor thing: can you add a comment to this compound condition? The meaning is not totally obvious. Have feeling (probably a wrong one) that some brackets are missing.

**Review by FoggyFinder — APPROVED** [review]
I've left one comment can you take a look on it? Anyway, approved.

File: CSharpMath.Forms/MathInputButton.cs
Comment (SymboLinker) [reply CSharpMath.Forms/MathInputButton.cs:48]: I totally agree that that piece of code is not very good readable. However, I'm not a fan of comments in code. Maybe I should refactor it? This could be commented code:
```
// unicode "\u25A1" is converted into latex "\square". Why only take this into account for the resting placeholder? I guess that the active placeholder will never be an empty square.
            var restingNucleus = LaTeXSettings.PlaceholderRestingNucleus.Replace("\u25A1", @"\square");
            var coloredPlaceholderRestingNucleus = LatexHelper.SetColor(restingNucleus, restingPlaceholderColor);
            var coloredPlaceholderActiveNucleus = LatexHelper.SetColor(LaTeXSettings.PlaceholderActiveNucleus, activePlaceholderColor);
            if (Input == MathKeyboardInput.Power || Input == MathKeyboardInput.Subscript) {
// In case of a power or subscript, the placeholder on the RIGHT is active on first appearance.
              latex = latex.ReplaceFirstOccurrence(restingNucleus, coloredPlaceholderRestingNucleus);
              latex = latex.ReplaceLastOccurrence(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
            } else if (
// in ALL other cases of two placeholders,
// with the same nucleus (both a filled square, but with a different color)
LaTeXSettings.PlaceholderRestingNucleus == LaTeXSettings.PlaceholderActiveNucleus && latex.SubStringCount(LaTeXSettings.PlaceholderActiveNucleus) > 1
// or with a different nucleus, a filled square vs an empty square
                 || restingNucleus != LaTeXSettings.PlaceholderActiveNucleus && latex.Contains(restingNucleus) && latex.Contains(LaTeXSettings.PlaceholderActiveNucleus)) {
// the placeholder on the LEFT should be active on first appearance
              latex = latex.ReplaceFirstOccurrence(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
              latex = latex.ReplaceLastOccurrence(restingNucleus, coloredPlaceholderRestingNucleus);
            } else {
// single placeholder should be active (let's not overwrite, but set the right color in one go)
              latex = latex.Replace(LaTeXSettings.PlaceholderActiveNucleus, coloredPlaceholderActiveNucleus);
            }
```
I realize that refactoring the code with same functional results, but wrong latex, is possible. I will add another unit test.

Comment (FoggyFinder) [reply CSharpMath.Forms/MathInputButton.cs:48]: Good enough for me. Yeah, I think refactoring makes sense but it can be done it later, not in this PR.

**Comment by SymboLinker** [comment]
Thanks @Happypig375 and @FoggyFinder!

**Pull request merged**
